R/find_exports.R

```diff
@@ -2,18 +2,10 @@ find_exports <- function(clean_lns) {
   ids <- find_extendr_attrs_ids(clean_lns)
   start <- ids
   end <- dplyr::lead(ids, default = length(clean_lns) + 1L) - 1L
-  purrr::map2_dfr(
-    start,
-    end,
-    ~ extract_meta(clean_lns[seq(.x, .y, by = 1L)])
-  ) %>%
-    # Keeps only name, type (fn|impl) and lifetime of impl
-    # if present.
-    dplyr::transmute(
-      .data$name,
-      type = dplyr::if_else(is.na(.data$impl), "fn", "impl"),
-      .data$lifetime
-    )
+
+  purrr::map2_dfr(start, end, ~ extract_meta(clean_lns[.x:.y])) %>%
+    dplyr::mutate(type = dplyr::coalesce(.data$impl, .data$fn)) %>%
```

Comment:
While we are at it, here is an issue:
``` r
 rextendr::rust_source(code = "fn test() {}")
#> i build directory: 'C:/Users/.../AppData/Local/Temp/RtmpyoVw1l/file55b864127a'
#> Error in `dplyr::mutate()`:
#> i In argument: `type = dplyr::coalesce(.data$impl, .data$fn)`.
#> Caused by error in `.data$impl`:
#> ! Column `impl` not found in `.data`.
#> Backtrace:
#>      x
#>   1. +-rextendr::rust_source(code = "fn test() {}")
#>   2. | \-rextendr::make_module_macro(code, module_name)
#>   3. |   \-rextendr:::find_exports(sanitize_rust_code(lines))
#>   4. |     \-... %>% ...
#>   5. +-dplyr::select(., dplyr::all_of(c("name", "type", "lifetime")))
#>   6. +-dplyr::mutate(., type = dplyr::coalesce(.data$impl, .data$fn))
#>   7. +-dplyr:::mutate.data.frame(...)
#>   8. | \-dplyr:::mutate_cols(.data, dplyr_quosures(...), by)
#>   9. |   +-base::withCallingHandlers(...)
#>  10. |   \-dplyr:::mutate_col(dots[[i]], data, mask, new_columns)
#>  11. |     \-mask$eval_all_mutate(quo)
#>  12. |       \-dplyr (local) eval()
#>  13. +-dplyr::coalesce(.data$impl, .data$fn)
#>  14. | \-rlang::list2(...)
#>  15. +-impl
#>  16. +-rlang:::`$.rlang_data_pronoun`(.data, impl)
#>  17. | \-rlang:::data_pronoun_get(...)
#>  18. \-rlang:::abort_data_pronoun(x, call = y)
#>  19.   \-rlang::abort(msg, "rlang_error_data_pronoun_not_found", call = call)
```

<sup>Created on 2023-04-19 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Just FYI. I will open a separate issue for this.

Reply:
Good note. `extract_meta()` should always return a df with the required columns even if 0 rows.

Reply:
See https://github.com/extendr/rextendr/issues/264